### ISSUE TITLE
Rem and em sizes are relatively different

### ISSUE THREAD

**Issue by msecret** [uswds/uswds#992]
The CSS guidelines states that `rem` units should be used for most situations unless a component needs to have units relative to a parent font size. Currently on the standards right now, using an rem unit is a completely different size then an em unit, which is inconsistent and potentially confusing. This is happening because we set a font-size of 10px on `html`, which sets the actual size for `rem` relative to 10px and the body's font size is set to `1.7rem`, which sets the `em` actual size relative to 17px. This means everywhere in the document `1rem` = `10px` and `1em` = `17px`.

As a developer, I'd expect 1rem to equal 1em in my elements _unless_ I specifically choose to change the em actual size by setting font size on a parent (like I did with an example below).

For an example of an element that should use em units view https://github.com/18F/cg-style/blob/master/src/css/components/hexagon.scss. This hexagon code can't use `%` units because `%` don't work in border. Instead, it uses `em`s and so that the size is relative to it's parents font-size.

@harrisj discussed his similar problems with this in the frontend channel. On cg-style, I reset both html and body to be 16px, the browser default (where did we even get 17px from??) and can use rems for everything besides that hexagon code.

**Comment by maya** [comment]
>where did we even get 17px from??

This was from the visual design of the type system when Source Sans Pro is the body font. You'll see when Merriweather is the body font, the type system changes, i.e. body for = 15px https://playbook.cio.gov/designstandards/visual-style/#typography.

How would we set the body font size to 17pm without effecting the 10px ems?

**Comment by carodew** [comment]
Am I wrong in thinking it is (or was) considered a best practice to set the baseline font size to 100%, so that users can zoom more easily? Sincere question – I thought that was the case but I'm certainly not putting effort into keeping up. 

But yeah, seems like the baseline em and rem should be the same in whatever scenario.

**Comment by maya** [comment]
@carodew setting it to 100% means that the fonts are set to whatever the user agent has - typically this is 16px. I don't think it's related to zooming. Setting it to 10px is a little trick developers use to ensure 1em = 10px so they can develop in a consistent way.

**Comment by msecret** [comment]
@carodew in some ways thats true, but I think it depends "how" users are zooming in or changing their browsers font size.

I found one article that says 100% should be used for when people set their font-size in their browser settings: http://terrillthompson.com/blog/589. I generally think this could warrant some more research as it depends on browser versions, etc.

There's also potential problems with the setting to 10px rule as talked about here: https://www.filamentgroup.com/lab/how-we-learned-to-leave-body-font-size-alone.html.

**Comment by maya** [comment]
Can you can set the parent element font-size to 1rem to get around that? I don't know that I see it is a problem that 1rem != 1em. 

If we didn't set the 10px on html, then 1em would equal the user agent size which would be 16px unless the user had a custom size.

**Comment by msecret** [comment]
What parent element do you mean? 

Also, was there a specific reason we didn't use `rem`s everywhere?

```
src/stylesheets/components/_alerts.scss:  margin-top: 1.5em;
src/stylesheets/components/_alerts.scss:  padding: 1em;
src/stylesheets/components/_footer.scss:      line-height: 2em;
src/stylesheets/components/_footer.scss:    padding-left: 1em;
src/stylesheets/components/_forms.scss:    margin-bottom: 1.5em;
src/stylesheets/components/_forms.scss:      padding-left: 2.7em;
src/stylesheets/components/_forms.scss:      padding-right: 2.7em;
src/stylesheets/core/_utilities.scss:      margin-right: 2em;
src/stylesheets/core/_utilities.scss:      margin-top: 0.5em;
...
```

And 
```
If we didn't set the 10px on html, then 1em would equal the user agent size which would be 16px unless the user had a custom size.
```
em would be 16px for the default then relative to 17rem set on `<body>`, so it wouldn't be 16px, but something else which I don't feel like doing the math for right now.

Another option here is rather then setting our body to 17px in any way, we set it on the elements we need it for, or the `*` selector.

So rather then this:
```
html { font-size: 10px; }
body { font-size: 1.7rem; }
```

We do this:
```
html { font-size: 100%; }
p, span, label, li, ... { font-size: 1.7rem; }
// I'm actually not sure about this next one, it might cause the same problems.
* { font-size: 1.7rem }
```

Another option is to set html to 17px, and have everything be relative to 17px. This would potentially go against both best practices mentioned in those articles.

**Comment by maya** [comment]
So are we moving away from the 10px base font size? What about this?

```
html { font-size: 10px; }
p, span, label, li, ... { font-size: 1.7rem; }
```

**Comment by maya** [comment]
>Also, was there a specific reason we didn't use rems everywhere?

I think we'd need to go in and check to make sure ems and rems are being used appropriately. For example, we use ems on heading and paragraph text margins to ensure that they scale with the text size. We'd have to go one by one and recheck the others.

**Comment by Drakenhart** [comment]
Setting a base font size is needed for responsive design if you want to control your typography more consistently.

**Comment by msecret** [comment]
I just read an article that did a great job discussing the tradeoffs in what we're discussing:

http://zellwk.com/blog/rem-vs-em/

One of it's primary points is 

"Although this is possible, it’s often a bad idea to set font-size in the <html> to a pixel value because it overrides the user’s browser settings."

I've come to agree with this statement. Sure, it means that you're typography may not always be as consistent but trying to control you typography could end up hindering certain users of your page that have specific needs in terms of their font usage. While on some sites I'd recommend setting a px font-size on html for consistency, I think the standards should not due to it's commitment to accessibility. These means our `html { font-size: 10px; }` rule should be removed. I still think it's fine to set a percentage value on `html` but that is also a different discussion on how to setup the typography of the page and where `em` and `rem` units should be the same size by default.

**Comment by carodew** [comment]
:100:

**Comment by maya** [comment]
@msecret using this method, would we still want to set our paragraph text to what the designers outlined? So (17÷16)em or 1.0625em? Or would we set our HTML or body to 106.25% and our paragraph text to 1em?

**Comment by msecret** [comment]
@maya that's more about whether we want `em`s and `rem`s to equal the same value. I personally would want them to, which would mean setting p to 1.0625rem. But I'd like to get some insight from other developers on what they'd expect from a framework.

**Comment by msecret** [comment]
@maya the other option is to set a certain percentage on `html` so that the base `em` and `rem` sizes are relative to 17px everywhere. I haven't fully considered which option is the best, both have trade-offs.

**Comment by maya** [comment]
The one downside of the first approach is that adding text without a defined font size is that the text will appear tiny. Setting a percentage on `html` ensures a little more stability in our font sizes across elements and components.

**Comment by Drakenhart** [comment]
Typography on the web has been a favorite topic for me for years. There are pros and cons to bother sides of the argument.

Setting strict values and not letting the user control them is a big problem. This is poor usability and poor accessibility.

 But setting a base (preferably not pixel based, you can do it as a percentage too), and then adjusting it by ems and rems gives you more flexibility and has been the compromise set between strict control (preferred by those who do print graphic design work) and no font control (preferred by those who understand the nature of the web and don't care about what happens to their fonts).

I come from that hybrid theory stance, from the no font control stance. This is due to accessibility concerns in regards to readability of text. The cpl (character per line) is in direct relation to the font size and line height. A bigger font means a wider cpl, and to do that you need to declare a base font size to sort out the rest of the Web Typography math.

So even if a user scales it up, it should responsively change to fit and not deform the site. That's the key to the compromise. Not deforming the site, but still have it readable (and user controllable).

^.^
http://typecast.com/blog/a-more-modern-scale-for-web-typography

http://alistapart.com/article/more-meaningful-typography
http://alistapart.com/column/responsive-typography-is-a-physical-discipline
https://bootstrapbay.com/blog/web-typography-best-practices/
https://www.smashingmagazine.com/2013/05/typographic-design-patterns-practices-case-study-2013/

**Comment by kaelig** [comment]
I'd like to mention that if you're planning to set the font-size on the HTML element to `%` instead of `px`, you might also want to apply this hack for IE11:

```css
/**
 * 1. Set baseline font size to 10px
 * 2. Fixes an IE11 bug where rounding with % would create scaling issues  
 *    See http://bit.ly/1g4X0bX
 * 4. Set default text size to 17px (at default zoom level)
 */
html {
    font-size: 62.5%; /* 1 */
    font-size: calc(1em * .625); /* 2 */
}
body {
    font-size: 1.7em; /* 3 */
}
```

This trick used to be in place on theguardian.com, see source: https://github.com/guardian/guss-typography/blob/v3.0.0/_typography.mixins.scss – Note that they switched to `font-size: 100%` on the html element (and use PostCSS to convert px to rem).

**Comment by btn0s** [comment]
Is there a consensus on this? Hasn't been closed but given it was over a year ago I'd imagine it's been resolved.

**Comment by maya** [comment]
@btn0s not yet, but it's on the [roadmap](https://standards.usa.gov/whats-new/product-roadmap/). Stay tuned.

**Comment by btn0s** [comment]
@maya Awesome! Just looking around for things to do in here so I was wondering what the status was. Thanks!

**Comment by maya** [comment]
@btn0s Thanks for checking in! 

We have a bunch bugs that could always use help: https://github.com/uswds/uswds/issues?q=is%3Aissue+is%3Aopen+label%3A%22%5BType%5D+Bug%22 - you could also look by high priority labeled issues too. 

Feel free to email us or join our Slack group if you want to chat more https://chat.18f.gov/ 😄

**Comment by maya** [comment]
Base font size will be configurable in version 2.0. Closing for now.
